### ISSUE TITLE
Long URLs not tweet-able

### ISSUE THREAD

**Issue by hanjo** [umputun/rss2twitter#11]
Hi,

I noticed an issue when both title and URL are quite long.

Let's assume title is 200 chars and URL is 100 chars. This would normally work, since the Twitter API is reducing the URL to 23 chars, and the resulting 200 + 3 (" - ") + 23 chars is way less then the maximum of 280.

The app however removes the URL, since it traces back from the end (300>275) to find the first space, which is just after the dash and removes everything from there, leaving just the title and the dash.

Any idea how to fix this?

Thanks!

**Comment by umputun** [comment]
A real example would be helpful, but I think i understand what you mean - this [format func](https://github.com/umputun/rss2twitter/blob/master/app/main.go#L126) treats the text without any consideration about urls and will strip a part of url (or even the whole url) the same way it strips any regular words.

Probably the solution is to make format smarter and  url-aware. i.e. instead of going back till the first space it should split the text first and process each "word" from end to start without striping url if possible. Counting should be adjusted as well because reduced URL should be handled in total len calculation differently.

If you want to try it - all yours.

**Comment by hanjo** [comment]
Ok, let me give you an example: Suppose I want to tweet:
`Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126`
my expection is that the tweet will be
`Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores https://t.co/something`
but in fact it will become
`Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores ...`
because that is how the format function will truncate the input.

Actually I don't think that the format function is the right spot to fix the issue. Recognizing a URL in the same way as twitter does is quite complex. Check this for how twitter-text does it:
https://github.com/twitter/twitter-text/blob/master/java/src/main/java/com/twitter/twittertext/Regex.java

I think a better way would be to do this at template level. Unfortunately, I've no experience in Go, but suppose the format function wasn't there, a simple way would be to use a template such as `{{ printf \"%.250s\" .Title }} {{.Link}}`. Of course, this is not as smart as the format function, but it would reliably truncate the text after 250 characters, leaving enough space for the URL.

According to the Go documentation, it should also be possible to execute a custom function, so instead of the simple printf function it should be possible to use the existing format function.
Then, [Line 117](https://github.com/umputun/rss2twitter/blob/6638bf54a62d0130150f635e21ebafc45bf75aa0/app/main.go#L117) just needs to have the format function removed, e.g.
```go
return strings.Replace(b1.String(), `\n`, "\n", -1) // \n in template
```

What do you think?

**Comment by hanjo** [comment]
Ok, I researched a little bit on the topic, and I believe a solution of this would be the below code, but since I do not know any Go, I'm not sure if it is correct. Would appreciate your review :-)
```go
// do runs event loop getting rss events and publishing them
func do(ctx context.Context, notif notifier, pub publisher.Interface, tmpl string) {
	log.Printf("[INFO] message template - %q", tmpl)
	ch := notif.Go(ctx)
	for event := range ch {
		err := pub.Publish(event, func(r rss.Event) string {
			b1 := bytes.Buffer{}
			if err := template.Must(template.New("twi").Funcs(funcMap).Parse(tmpl)).Execute(&b1, event); err != nil { //nolint
				// template failed to parse record, backup predefined format
				return fmt.Sprintf("%s - %s", format(r.Title, 250), r.Link)
			}
			return strings.Replace(b1.String(), `\n`, "\n", -1) // \n in template
		})
		if err != nil {
			log.Printf("[WARN] failed to publish, %s", err)
		}
	}
}

// Map name format to format function
var funcMap = template.FuncMap{
	"format": format,
}
```

With this, it should be possible to use a template such as `{{format .Title 250}} {{.Link}}`

**Comment by umputun** [comment]
I'm a little bit confused because I'm not sure where the link you expected to see is coming from. Is it rss item link, or a part of the text/title? I'm also not sure what " Suppose I want to tweet ..." means in this context. This app tweets from rss items and the better example indicating the actual problem will be the real rss item, i.e. something like this:

```xml
		<item>
			<title>
				Радио-Т 625
			</title>
			<description>
				<![CDATA[
				<p><img src="https://radio-t.com/images/radio-t/rt625.jpg" alt=""></p>
				<ul>
				<li><a href="https://www.cnet.com/news/logitech-reportedly-in-talks-to-acquire-plantronics/">Logitech приобретает Plantronics</a>.</li>
				<li><a href="https://bossasaservice.life">Boss as a Service</a>.</li>
				<li><a href="https://gcemetery.co/">Кладбище проектов Google</a>.</li>
				<li><a href="https://developer.github.com/actions/creating-workflows/creating-a-new-workflow/">Actions в реальной жизни</a>.</li>
				<li><a href="https://www.forbes.com/sites/janakirammsv/2018/11/21/aws-adds-the-most-anticipated-feature-to-amazon-ec2/">AWS добавила ума EC2</a>.</li>
				<li><a href="https://habr.com/post/427211/">Electron это Flash для десктопа</a>.</li>
				<li><a href="http://blairreeves.me/2018/11/09/dont-work-remotely/?resubmit=hn">Как решить вопрос удаленной работы</a>.</li>
				<li><a href="https://mashable.com/article/amazon-customer-data-leak/">Amazon потеряла данные пользователей</a>.</li>
				<li>Темы наших слушателей</li>
				</ul>
				<p><em>Спонсор этого выпуска <a href="https://www.digitalocean.com">DigitalOcean</a></em></p>
				<p><a href="http://cdn.radio-t.com/rt_podcast625.mp3">аудио</a> ● <a href="http://chat.radio-t.com/logs/radio-t-625.html">лог чата</a></p>
				<audio src="http://cdn.radio-t.com/rt_podcast625.mp3" preload="none"></audio>
				]]>
			</description>
			<link>
				https://radio-t.com/p/2018/11/24/podcast-625/
			</link>
			<guid>
				https://radio-t.com/p/2018/11/24//podcast-625/
			</guid>
			<pubDate>
				Sat, 24 Nov 2018 17:59:34 EST
			</pubDate>
			<author>
				Umputun, Bobuk, Gray, Ksenks
			</author>
			<summary>
				<![CDATA[
				<p><img src="https://radio-t.com/images/radio-t/rt625.jpg" alt=""></p>
				<ul>
				<li><a href="https://www.cnet.com/news/logitech-reportedly-in-talks-to-acquire-plantronics/">Logitech приобретает Plantronics</a>.</li>
				<li><a href="https://bossasaservice.life">Boss as a Service</a>.</li>
				<li><a href="https://gcemetery.co/">Кладбище проектов Google</a>.</li>
				<li><a href="https://developer.github.com/actions/creating-workflows/creating-a-new-workflow/">Actions в реальной жизни</a>.</li>
				<li><a href="https://www.forbes.com/sites/janakirammsv/2018/11/21/aws-adds-the-most-anticipated-feature-to-amazon-ec2/">AWS добавила ума EC2</a>.</li>
				<li><a href="https://habr.com/post/427211/">Electron это Flash для десктопа</a>.</li>
				<li><a href="http://blairreeves.me/2018/11/09/dont-work-remotely/?resubmit=hn">Как решить вопрос удаленной работы</a>.</li>
				<li><a href="https://mashable.com/article/amazon-customer-data-leak/">Amazon потеряла данные пользователей</a>.</li>
				<li>Темы наших слушателей</li>
				</ul>
				<p><em>Спонсор этого выпуска <a href="https://www.digitalocean.com">DigitalOcean</a></em></p>
				<p><a href="http://cdn.radio-t.com/rt_podcast625.mp3">аудио</a> ● <a href="http://chat.radio-t.com/logs/radio-t-625.html">лог чата</a></p>
				<audio src="http://cdn.radio-t.com/rt_podcast625.mp3" preload="none"></audio>
				]]>
			</summary>
			<image href="https://radio-t.com/images/radio-t/rt625.jpg" />
			<enclosure url="http://cdn.radio-t.com/rt_podcast625.mp3" type="audio/mp3" length="90301564" />
			<author>
				podcast@radio-t.com (Umputun, Bobuk, Gray, Ksenks)
			</author>
			<content url="http://cdn.radio-t.com/rt_podcast625.mp3" fileSize="90301564" type="audio/mp3" />
			<explicit>
				no
			</explicit>
			<subtitle>
				Подкаст выходного дня - импровизации на темы высоких технологий
			</subtitle>
			<keywords>
				hitech,russian,radiot,tech,news,радио
			</keywords>
		</item>
```

If you can provide both RSS item in question and your template, it will help to clarify the issue.

**Comment by hanjo** [comment]
Well, I didn't mean "I" want to tweet, I meant the RSS of course. Sorry for the confusion.

I don't have a real world example right now, because the feed is changing so fast. Maybe I can get one tomorrow, but when using the example above, it would be like so:
```xml
<item>
	<title>
		Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores
	</title>
	<link>
		https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126
	</link>
	<description>
		Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
	</description>
	<guid>
		https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126
	</guid>
</item>
```

The template is as simple as `{{.Title}} {{.Link}}`

**Comment by umputun** [comment]
ok, got it.

Regarding your proposal - I'm not sure this version with template function is something ... user friendly. I mean, the syntax for templates is somewhat unusual for the regular users even now, but at least approachable. However, explaining to the non-techy user what this magic means (i.e. `{{format .Title 250}} {{.Link}}`) will be a challenge. And it gets worse if `.Text`involved too. Here mapping of individual elements will be more problematic. I mean for the template like this `{{.Title}} {{.Text}} - {{.Link}}`

From another hand, all we care about is limiting the overall size of the tweet, counting `.Link` length in a special way, and trimming other parts of the tweet without touching the link. Currently, I don't see a simple and elegant way to do it. It gets kind of tricky as soon as we need to address {{.Link}} not just in the end but also in the middle.

a partial approach can be to use [t.co](https://t.co/) internally, so the current logic will work better. Not sure if t.co has an api for this.

**Comment by hanjo** [comment]
Ok, you have a point there. I have a different proposal then. The following function should do what you had originally suggested above: essentially, it leaves all the URLs in place (wherever they are) and just reduces the text if necessary.
It has a couple of limitations though:
1. Due to the way it works, it removes any consecutive whitespace. Mutliple whitespaces will be reduced to a single whitespace
2. It iterates over the string twice, maybe that can be avoided, but I couldn't see how, since I need to count the number of URLs first, before I can start with the actuall process
3. The regex to identify a URL is simplified and not as sophisticated as the one twitter-text uses
4. Please keep in mind that I have just started programming Go yesterday, so this is most likely not very good code
```go
var urlregex = `https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)`
var urllength = 23

// format cleans text (removes html tags) and shrinks result
func format(inp string, max int) string {
	res := striphtmltags.StripTags(inp)

	// splits the string int an array at each whitespace
	var words = strings.Fields(res)
	// creates a boolean array of the same length, used later to store which word is an URL
	var isURL = make([]bool, len(words))
	// count of the URLs in the string
	var urls = 0

	// iterate through the words, identify the URLs and store the count of URLs
	for i, v := range words {
		matched, _ := regexp.MatchString(urlregex, v)
		isURL[i] = matched
		if matched {
			urls++
		}
	}

	// start with an empty string for the output
	res = ""
	// reduce the required space for all the URLs from the total of available length
	max -= urls * (urllength + 1) // the +1 is for the space

	// define what's to appear when the length limit is hit
	var end = " ..."
	// iterate through the words and cut everything that is over the limit
	for i, v := range words {
		// helper to identify if we need to add a space (which is always true except for the very beginning)
		var addspace = 0
		if len(res) > 0 {
			addspace = 1
		}

		if len(res)+addspace+len(v) < max-len(end) { // if there is enough space left, add the word to the output
			if addspace > 0 {
				res += " "
			}
			res += v
		} else if !isURL[i] { // if not, replace the word with the "end" string (and set it to an empty string so the next words are just removed)
			res += end
			end = ""
		} else if isURL[i] { // if it's a URL, add it to the output
			if addspace > 0 {
				res += " "
			}
			res += v
		}
	}
	return res
}
```
you need to call it using `format(someString, 280)` (not 275), since it calculates the actual space required automatically.

You can try some examples here: https://go.dev/play/p/_6aJOCmykPq

Let me know if this will work.

**Comment by umputun** [comment]
thx for the snippet. I think it can be significantly simplified if we replace the format function with something getting a template and the original struct and producing the final, limited len string, i.e. `format(r rss.Event, template string, max int) string` 

in this case no need to regex anything as we kind of know each element already. This way all we need to do is to remove {{.Link}} from the template first (`strings.Replace(template, {{.Link}}, "", -1)`), apply it, which will give us the len of the rest of things. From here we can trim if needed and reapply the original template. Well, have not tried it myself, but I think it will work

Feel free to submit a PR

**Comment by umputun** [comment]
pls, check out the master image and let me know how it worked for you. It should handle links properly and support smarter trim without affecting links or the constant (predefined) parts of the template. I have tried to address as many corner cases as I can think about, but feel free to offer more tests [here](https://github.com/umputun/rss2twitter/blob/master/app/main_test.go#L124)

**Comment by hanjo** [comment]
Thank you so much, I was still considering how to best implement what you've suggested.

I tested your code, and generally it seems to be working, but there seem to be two issues still. Consider the following case:

You have a title that is 300 chars and a link that is 100 chars. If your template is `{{.Title}} - {{.Link}}` the output will be approximately 255 chars ending in "dolor sit amet, consetetur...  -... ". My point is, even though there is enough space left for the URL, it is replaced with "... ". I guess this is due to calling `trimWithDots` on the return statement of the `formatMsg` function.

When you do it the other way around with the template `{{.Link}} - {{.Title}}` the output will start with the link just fine, but it will cut the title after ~178 characters, because it was counting the entire length of the link, not the `shortLinkLen`.

I used the following for my test:
```go
	var ev1 = Event{}
	ev1.Title = "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 sed diam voluptua. At vero eos et accusam et justo duo dolores"
	ev1.Link = "https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126"
	var ev2 = Event{}
	ev2.Title = "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores"
	ev2.Link = "https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126"
	var ev3 = Event{}
	ev3.Title = "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores"
	ev3.Link = "https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126"

	fmt.Println(formatMsg(ev1, "{{.Title}} - {{.Link}}", 279))
	fmt.Println(formatMsg(ev2, "{{.Title}} - {{.Link}}", 279))
	fmt.Println(formatMsg(ev3, "{{.Title}} - {{.Link}}", 279))
	fmt.Println(formatMsg(ev1, "{{.Link}} - {{.Title}}", 279))
	fmt.Println(formatMsg(ev2, "{{.Link}} - {{.Title}}", 279))
	fmt.Println(formatMsg(ev3, "{{.Link}} - {{.Title}}", 279))

```
This is the output:
```
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur...  -... 
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur...  -... 
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores -... 
https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 - Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et... 
https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 - Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et... 
https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 - Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et... 
```

**Comment by umputun** [comment]
thx, this is very useful input. I have added your examples to my tests and adjusted formatter by moving the final trim to the proper place (only if we do the backup predefined formatting)

**Comment by hanjo** [comment]
Thanks! This is really very good now, see the test results here:
```
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur...  - https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur...  - https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores - https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126
https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 - Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur... 
https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 - Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores       Lorem ipsum dolor sit amet, consetetur... 
https://github.com/umputun/rss2twitter/blob/d5c89112e4eb8ed8d1d0717526804bc145202fe5/app/main.go#L126 - Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores
```
Just as it should be :-) 
I'd say, this is ready to be released as 0.7.0, don't you think?

**Comment by umputun** [comment]
yep, will tag it shortly
